File: cdk/typescript/rds-aurora-serverlessv2/lib/rds-aurora-serverlessv2-stack.ts
import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { 
  Vpc,
  SubnetType,
  IpAddresses,
} from 'aws-cdk-lib/aws-ec2';
import * as rds from 'aws-cdk-lib/aws-rds';
import { Key } from 'aws-cdk-lib/aws-kms';
import { HostedRotation } from 'aws-cdk-lib/aws-secretsmanager';

export class RdsAuroraServerlessv2Stack extends cdk.Stack {
  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Create private VPC with cidr block 10.0.0.0/16
    const vpc = new Vpc(this, 'RdsVpc', {
      ipAddresses: IpAddresses.cidr('10.0.0.0/16'),
      restrictDefaultSecurityGroup: false,
      natGateways: 0,
      subnetConfiguration: [
        {
          cidrMask: 24,
          name: 'private',
          subnetType: SubnetType.PRIVATE_ISOLATED,
        }
      ]
    });

    // Create username and password secret for DB Cluster
    const secret = new rds.DatabaseSecret(this, 'RdsSecret', {
      username: 'admin',
      encryptionKey: new Key(this, "DBEncryptionKey", {
        enableKeyRotation: true,
      }),
    });

    // Add Rotation schedule to DB Cluster secret
    secret.addRotationSchedule('RotationSchedule', {
      automaticallyAfter: cdk.Duration.days(30),
      hostedRotation: HostedRotation.mysqlSingleUser(),
    });

    // Define a serverless Aurora database cluster writer instance
    const writer = rds.ClusterInstance.serverlessV2( "writer", {
      autoMinorVersionUpgrade: true,
    });

    // Define a serverless Aurora database cluster reader instance that scales with writer nodes
    const reader = rds.ClusterInstance.serverlessV2( "reader", {
      scaleWithWriter: true,
      autoMinorVersionUpgrade: true,
    });

    // Create Aurora Serverless V2 MySQL cluster with version 3.04.0 that passes security and linting checks
    const cluster = new rds.DatabaseCluster(this, 'AuroraDatabaseCluster', {
      engine: rds.DatabaseClusterEngine.auroraMysql({ version: rds.AuroraMysqlEngineVersion.VER_3_04_0 }),
      writer: writer,
      readers: [
        reader,
      ],
      serverlessV2MinCapacity: 0.5,
      serverlessV2MaxCapacity: 2,
      vpc: vpc,
      vpcSubnets: { subnetType: SubnetType.PRIVATE_ISOLATED },
      credentials: rds.Credentials.fromSecret(secret),
      iamAuthentication: true,
      backtrackWindow: cdk.Duration.seconds(30),
      monitoringInterval: cdk.Duration.seconds(5),
      storageEncrypted: true,
    });

    // Suppress CloudFormation Guard checks that are inherited from the Aurora Cluster and 
    // do not apply to individual Aurora Cluster Instances.
    const cfnWriter = cluster.node.findChild("writer").node.defaultChild as rds.CfnDBInstance;
    cfnWriter.cfnOptions.metadata = {
      guard: {
        SuppressedRules: [
          "DB_INSTANCE_BACKUP_ENABLED",
          "RDS_STORAGE_ENCRYPTED",
          "RDS_INSTANCE_LOGGING_ENABLED",
          "RDS_INSTANCE_DELETION_PROTECTION_ENABLED",
          "RDS_SNAPSHOT_ENCRYPTED",
          "RDS_MULTI_AZ_SUPPORT",
        ],
      }
    };

    // Suppress CloudFormation Guard checks that are inherited from the Aurora Cluster and 
    // do not apply to individual Aurora Cluster Instances.
    const cfnReader = cluster.node.findChild("reader").node.defaultChild as rds.CfnDBInstance;
    cfnReader.cfnOptions.metadata = {
      guard: {
        SuppressedRules: [
          "DB_INSTANCE_BACKUP_ENABLED",
          "RDS_STORAGE_ENCRYPTED",
          "RDS_INSTANCE_LOGGING_ENABLED",
          "RDS_INSTANCE_DELETION_PROTECTION_ENABLED",
          "RDS_SNAPSHOT_ENCRYPTED",
          "RDS_MULTI_AZ_SUPPORT",
        ],
      }
    };

  }
}

File: cdk/python/rds-aurora-serverlessv2/rds_aurora_serverlessv2/rds_aurora_serverlessv2_stack.py
from aws_cdk import (
    Duration,
    Stack,
    aws_rds as rds,
    aws_ec2 as ec2,
    aws_kms as kms,
)
from constructs import Construct

class RdsAuroraServerlessv2Stack(Stack):

    def __init__(self, scope: Construct, construct_id: str, **kwargs) -> None:
        super().__init__(scope, construct_id, **kwargs)

        # Create basic VPC with cidr block 10.0.0.0/16
        vpc = ec2.Vpc(
            self,
            "VPC",
            ip_addresses=ec2.IpAddresses.cidr("10.0.0.0/16"),
            restrict_default_security_group=False,
            max_azs=2,
            nat_gateways=0,
            subnet_configuration=[
                ec2.SubnetConfiguration(
                    subnet_type=ec2.SubnetType.PRIVATE_ISOLATED,
                    cidr_mask=24,
                    name="RDS"
                ),
            ],
        )

        # Create username and password secret for RDS instance
        rdsusernamesecret = rds.DatabaseSecret(
            self,
            "RDSSecret",
            username="admin",
            encryption_key=kms.Key(
                self,
                "MySQL-EncryptionKey",
                enable_key_rotation=True,
                ),
        )

        # Create Aurora cluster serverless writer instance
        aurora_writer=rds.ClusterInstance.serverless_v2(
            "writer",
            auto_minor_version_upgrade=True,
        )

        # Create Aurora cluster serverless reader instance
        aurora_readers=[
            rds.ClusterInstance.serverless_v2(
            "reader",
            scale_with_writer=True,
            auto_minor_version_upgrade=True,
            ),
        ]

        # Create Aurora MySQL 3.4.0 Serverless v2 cluster that passes security and linting checks
        aurora_cluster = rds.DatabaseCluster(
            self,
            "AuroraCluster",
            engine=rds.DatabaseClusterEngine.aurora_mysql(version=rds.AuroraMysqlEngineVersion.VER_3_04_0),
            writer=aurora_writer,
            readers=aurora_readers,
            serverless_v2_min_capacity=0.5,
            serverless_v2_max_capacity=2,
            backup=rds.BackupProps(
                retention=Duration.days(1),
            ),
            credentials=rds.Credentials.from_secret(rdsusernamesecret),
            vpc=vpc,
            vpc_subnets=ec2.SubnetSelection(subnet_type=ec2.SubnetType.PRIVATE_ISOLATED),
            backtrack_window=Duration.seconds(30),
            cloudwatch_logs_exports=["audit", "error", "general", "slowquery"],
            monitoring_interval=Duration.seconds(5),
            iam_authentication=True,
            storage_encrypted=True,
            storage_encryption_key=kms.Key(
                self,
                "AuroraStorageEncryptionKey",
                enable_key_rotation=True,
            ),
            deletion_protection=True,
        )

        rdsusernamesecret.node.default_child.cfn_options.metadata = {
            "guard": {
                "SuppressedRules": [
                    "SECRETSMANAGER_ROTATION_ENABLED_CHECK"
                ]
            }
        }

        # Suppress Cloudformation Guard settings that do not apply to Aurora Cluster Instances, inherited from
        # the Aurora Database Cluster Properties
        aurora_cluster.node.find_child("writer").node.default_child.cfn_options.metadata = {
            "guard": {
                "SuppressedRules": [
                    "DB_INSTANCE_BACKUP_ENABLED",
                    "RDS_STORAGE_ENCRYPTED",
                    "RDS_INSTANCE_LOGGING_ENABLED",
                    "RDS_INSTANCE_DELETION_PROTECTION_ENABLED",
                    "RDS_SNAPSHOT_ENCRYPTED",
                    "RDS_MULTI_AZ_SUPPORT",
                ]
            }
        }

        # Suppress Cloudformation Guard settings that do not apply to Aurora Cluster Instances, inherited from
        # the Aurora Database Cluster Properties
        aurora_cluster.node.find_child("reader").node.default_child.cfn_options.metadata = {
            "guard": {
                "SuppressedRules": [
                    "DB_INSTANCE_BACKUP_ENABLED",
                    "RDS_STORAGE_ENCRYPTED",
                    "RDS_INSTANCE_LOGGING_ENABLED",
                    "RDS_INSTANCE_DELETION_PROTECTION_ENABLED",
                    "RDS_SNAPSHOT_ENCRYPTED",
                    "RDS_MULTI_AZ_SUPPORT",
                ]
            }
        }


